File: BasicToAdvance/Python/Day-15/BSTSuccessor.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


def find_min(node):
    while node and node.left:
        node = node.left
    return node


def successor(root, key):
    succ = None
    curr = root

    while curr:
        if key < curr.val:
            succ = curr
            curr = curr.left
        elif key > curr.val:
            curr = curr.right
        else:
            if curr.right:
                return find_min(curr.right)
            break

    return succ


# Example usage
if __name__ == "__main__":
    root = Node(20)
    root.left = Node(10)
    root.right = Node(30)
    root.left.left = Node(5)
    root.left.right = Node(15)
    root.right.left = Node(25)
    root.right.right = Node(40)

    print(successor(root, 20))  # 25
    print(successor(root, 10))  # 15
    print(successor(root, 40))  # None
    print(successor(root, 5))   # 10

File: BasicToAdvance/Java/Day-15/BSTSuccessor.java
// Find the inorder successor of a given key in a Binary Search Tree.

class BSTSuccessor {
    static class Node {
        int val;
        Node left;
        Node right;

        Node(int val) {
            this.val = val;
        }
    }

    static Node findMin(Node node) {
        while (node != null && node.left != null) {
            node = node.left;
        }
        return node;
    }

    static Node successor(Node root, int key) {
        Node successor = null;

        while (root != null) {
            if (key < root.val) {
                successor = root;
                root = root.left;
            } else if (key > root.val) {
                root = root.right;
            } else {
                break;
            }
        }

        if (root != null && root.right != null) {
            return findMin(root.right);
        }

        return successor;
    }

    public static void main(String[] args) {
        Node root = new Node(20);
        root.left = new Node(10);
        root.right = new Node(30);
        root.left.left = new Node(5);
        root.left.right = new Node(15);
        root.right.left = new Node(25);
        root.right.right = new Node(35);

        int key = 20;
        Node result = successor(root, key);

        if (result != null) {
            System.out.println("Successor of " + key + " is: " + result.val);
        } else {
            System.out.println("No successor exists for " + key);
        }
    }
}